### PR TITLE
Modularize glvis-js and add npm package config

### PR DESCRIPTION
**Pull request by tomstitt**
These changes will let us publish `glvis-js` to npm so that we can add it as a dependency to `pyglvis`, specifically the `jupyter-glvis` npm module (which I'm going to rename `glvis-jupyter` for consistency).

Specifically this means we can just `require` and use:
```js
var glvis = require("glvis");

stream = ...;
var div_id = "a-div-for-glvis";
var canvas = glvis.setupCanvas(div, 1000, 1000);
glvis.displayStream(div_id, canvas.id, stream);
```

I've tested that this works with:

a) the examples *examples/basic.html* and *examples/requirejs.html* with the `file://` protocol
  - the first loads the libraries into the global space with `<script>`
  - the second loads the libraries in a scope with `require`
 
b) node, with `echo -e "var glvis = require('glvis');\nglvis.info();" > test.js && node test.js`
c) webpack, with the Jupyter widget installer in `pyglvis` after I removed its `libglvis.js` and added this to its dependencies

Let me know if there are any other tests y'all would like to see.

**IMPORTANT:** I haven't yet published this to npm but I plan to publish it as is under the name `glvis`, please let me know if you'd prefer another name. Please also look at *package.json* and let me know if we should modify anything else, like _description_, _keywords_, _license_, ...

Once this is merged I will update `pyglvis` and we can stop getting all those dependabot emails =D

I added [prettier](https://prettier.io) as a dev dependency and we should just use that going forward for formatting. 

### Major Issues:

- [ ] ~Memory Leak when loading new stream~ - delayed for a future PR, the main issue of this happening when streaming data from MFEM has been fixed
    - safari automatically reloads my page after it uses ~5GB
- [x] Current view is not preserved for new streams (because we just call `startVisualization` every time)
    - this may also cause a performance hit 

### Issues
- [ ] ~Multiple fields per app "overwrite" each other (for example, you see the first then the 2nd replaces it after .5 seconds [the delay])~ - delayed for a future PR

### Todo:
- [x] Document `prettier`
- [x] Document git-lfs requirements for `glvis.js`
- [x] Understand why the "Choose File" doesn't work in Safari (works in Firefox and Chrome)
- [ ] ~Add loading artifact for "Choose File" so it doesn't look like the page is frozen~ - delayed for a future PR

### After Merging:
- [ ] Publish to npm

### CODE REVIEW

**Comment by tomstitt** [comment]
I can't seem to include Max on this, neither in the Reviewers or with "@" in this comment, do y'all know why?

**Comment by tzanio** [comment]
Can you try again @tomstitt ?

File: package.json [hunk 1]
```diff
@@ -0,0 +1,27 @@
+{
+  "name": "glvis",
+  "version": "0.1.0",
+  "description": "A lightweight WebGL tool for accurate and flexible finite element visualization",
+  "license": "LGPL-2.1",
```
Comment (tzanio) [inline package.json:5]: BSD-3?

File: examples/keys.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+| a -  Displays/Hides the axes       |
+| A -  Turns antialiasing on/off     |
+| c -  Toggle colorbar and caption   |
+| C -  Change the main plot caption  |
+| e -  Displays/Hides the elements   |
+| E -  Toggle the elements in the CP |
+| f -  Smooth/Flat/discont. shading  |
+| g -  Toggle background             |
+| h -  Displays help menu            |
+| i -  Toggle cutting plane          |
```
Comment (tzanio) [inline examples/keys.js:14]:
```suggestion
| i -  Toggle cutting plane          |
| I -  Toggle cutting plane algorithm|
```

File: examples/keys.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+| S -  Take snapshot/Record a movie  |
+| t -  Cycle materials and lights    |
+| \ -  Set light source position     |
+| u/U  Move the level surface        |
+| v/V  Add/Delete a level surface    |
+| w/W  Move bdr elements up/down     |
+| x/X  Rotate clipping plane (phi)   |
+| y/Y  Rotate clipping plane (theta) |
+| z/Z  Translate clipping plane      |
+| Ctrl+p - Print to a PDF file       |
```
Comment (tzanio) [inline examples/keys.js:37]:
```suggestion
| Ctrl+o - Element ordering curve    |
| Ctrl+p - Print to a PDF file       |
```

**Review by tzanio — COMMENTED** [review]
A first pass of simple edits with a suggestion -- can we switch the license to BSD?

**Comment by tzanio** [comment]
(Future) Suggestions:

1️⃣  Change `examples/keys.js` depending on the input type, the keys differ for scalar/vector and 2d/3d data, see https://github.com/GLVis/glvis/blob/master/lib/vssolution.cpp#L49, https://github.com/GLVis/glvis/blob/master/lib/vssolution3d.cpp#L29, etc.

2️⃣  Is there a way to improve the I/O e.g. corresponding to key `F6`. Some of this could be done now with the JS console, but doesn't work quite right...

File: package.json [hunk 4]
```diff
@@ -0,0 +1,27 @@
+{
+  "name": "glvis",
+  "version": "0.1.0",
```
Comment (tzanio) [inline package.json:3]: Should we list here the version of GLVis from which glvis.js was built instead?

**Comment by tzanio** [comment]
I've said it before, I'll say it again -- this is just super cool stuff. Thank you @tomstitt !

**Comment by tzanio** [comment]
Here is one cool application -- use the examples as a GLVis replacement (e.g. in ATPESC lectures):

1. Open e.g. `glvis-js/examples/basic.html`

2. Replace `socketstream` with `ofstream` to `glvis-js/examples/example_stream.js` e.g. in MFEM's ex1.cpp

```diff
-      socketstream sol_sock(vishost, visport);
+      // socketstream sol_sock(vishost, visport);
+      ofstream sol_sock("../../glvis-js/examples/example_stream.js");
+      sol_sock << "var example_stream = `";
+
       sol_sock.precision(8);
       sol_sock << "solution\n" << mesh << x << flush;
+
+      sol_sock << "`;\n";
```

3. Run the example as usual

4. Reload the `examples/basic.html` page

**Comment by tomstitt** [comment]
@tzanio wrote in reference to _package.json_:
> Should we list here the version of GLVis from which glvis.js was built instead?

I don't think we should right now since we'll probably be making changes and adding features to the frontend, which would mess with our npm releases unless we just incremented the final version value, but that could be confusing since we may introduce big changes. I'm open to other opinions though.

**Comment by tzanio** [comment]
ok,  but it may be good to indicate somewhere what GLVis version was used for the pre-built library

**Review by tomstitt — COMMENTED** [review]
> ok, but it may be good to indicate somewhere what GLVis version was used for the pre-built library

Totally agree, it's embedded in the commits that update `glvis.js` but we can have it in the README/queryable from the browser. Maybe once we're happy with the state of things (a version "1") we can just start tracking GLVis's version. npm requires semver so that constrains us a little

**Comment by tomstitt** [comment]
> Here is one cool application -- use the examples as a GLVis replacement (e.g. in ATPESC lectures):
> 
> 1.

I don't think we can poll the filesystem for browser security reasons and some googling makes me think that is true but there are 2 other options I can think of:

1) Use WebSockets in the browser + a WebSocket wrapper around MFEM and just send the stream data as normal (I'm trying this now)
2) Run in a different WebAssembly setting, like a desktop node app, which has more flexibility with the filesystem

**Comment by tzanio** [comment]
@kanye-quest can you take a look at the following issues:

- Handle `g` correctly
- Make sure the rendering is high dpi

**Comment by kanye-quest** [comment]
The issues with the 'g' key should be fixed by GLVis/glvis#134.

**Comment by tzanio** [comment]
> The issues with the 'g' key should be fixed by [GLVis/glvis#134](https://github.com/GLVis/glvis/pull/134).

This fixes `g` for me, but now I don't see the text (e.g. at the corner of the axis). Am I doing something wrong?

**Comment by kanye-quest** [comment]
Text seems to display fine for me; when rebuilding the javascript library, is the font file present in the glvis directory?

**Comment by tzanio** [comment]
> Text seems to display fine for me; when rebuilding the javascript library, is the font file present in the glvis directory?

My bad, I had incorrect font file, https://github.com/GLVis/glvis-js/pull/1/commits/76cc746569b7710bea8662c3e8dbab254fe2bf96

**Comment by tzanio** [comment]
With this diff the JS version looks high dpi on my Mac, but it is not antialiased (and `A` doesn't work). Can this be fixed?

```diff
diff --git a/lib/aux_vis.cpp b/lib/aux_vis.cpp
index 7f07715..b26edb8 100644
--- a/lib/aux_vis.cpp
+++ b/lib/aux_vis.cpp
@@ -1607,11 +1607,7 @@ vector<string> fc_font_patterns =
    "Helvetica:style=Regular",
 };
 
-#ifdef __EMSCRIPTEN__
-constexpr int default_font_size = 14;
-#else
 constexpr int default_font_size = 12;
-#endif
 int font_size = default_font_size;
 
 GlVisFont glvis_font;
diff --git a/lib/sdl.cpp b/lib/sdl.cpp
index 4559a08..e4cbdba 100644
--- a/lib/sdl.cpp
+++ b/lib/sdl.cpp
@@ -174,9 +174,9 @@ bool SdlWindow::createWindow(const char * title, int x, int y, int w, int h,
 
    Uint32 win_flags = SDL_WINDOW_OPENGL;
    // Hide window until we adjust its size for high-dpi displays
-   win_flags |= SDL_WINDOW_HIDDEN;
+   win_flags |= SDL_WINDOW_ALLOW_HIGHDPI | SDL_WINDOW_HIDDEN;
 #ifndef __EMSCRIPTEN__
-   win_flags |= SDL_WINDOW_ALLOW_HIGHDPI | SDL_WINDOW_RESIZABLE;
+   win_flags |= SDL_WINDOW_RESIZABLE;
 #endif
    SDL_GL_SetAttribute( SDL_GL_DOUBLEBUFFER, 1);
    SDL_GL_SetAttribute( SDL_GL_DEPTH_SIZE, 24);
```

**Comment by kanye-quest** [comment]
It seems that while you can request antialiasing at context creation for WebGL, the browser is free to ignore that request, and that antialiasing isn't controllable via `glEnable/glDisable(GL_MULTISAMPLE)` as it is on the desktop.

Controllable antialiasing is available for WebGL 2, but in a bit of an indirect way: we can create a multisampled renderbuffer and blit that to the backbuffer.

**Comment by tzanio** [comment]
> Controllable antialiasing is available for WebGL 2, but in a bit of an indirect way: we can create a multisampled renderbuffer and blit that to the backbuffer.

Is that hard to try? 

Maybe the difference is visible only on Macs, but I'd like to fix it if possible...

**Comment by tomstitt** [comment]
I've added a `sendKey` method on the object we're using in our examples which should make it easy to build a gui for any functionality we want as buttons. I included an example for switching the color palette, please add more! We can clean up the looks later.

**Comment by tzanio** [comment]
@kanye-quest -- I pushed the high DPI changes from https://github.com/GLVis/glvis-js/pull/1#issuecomment-748398475 in d7c8b622409a833f808922da72451be50dea8079.

We can revert if there is a better solution.

**Comment by tomstitt** [comment]
We can now have references with links like `https://glvis.org/live?stream=../data/ex9.saved` for using non-default streams when the page is loaded. I'm not sure yet about non-local urls.

**Comment by tzanio** [comment]
@tomstitt, if I press `h` I see exactly the help I need in the new `Log` tab. Can that be used to auto-generate/update the `Help` tab?

**Comment by tzanio** [comment]
Some feedback from Slack here: https://mfem.slack.com/archives/CB98FH3U4/p1608764738459800?thread_ts=1608757601.458300&cid=CB98FH3U4

**Comment by tomstitt** [comment]
> @tomstitt, if I press `h` I see exactly the help I'd like in the new `Log` tab. Can that be used to auto-generate/update the `Help` tab?

See this PR: https://github.com/GLVis/glvis/pull/139

Once it's merged I can push the changes here that use it

**Comment by tomstitt** [comment]
I believe all the changes for this PR are complete, please look it over if you want and we'll hopefully merge later today!

**Review by tzanio — APPROVED** [review]
Awesome, awesome work! 🚀

**Pull request merged**